IMLS/imls_scraper.py

```diff
@@ -0,0 +1,109 @@
+from zipfile import ZipFile
+from io import BytesIO
+import pandas as pd
+import requests
+from bs4 import BeautifulSoup
+import pandas as pd
+import re
+from pymongo import MongoClient, errors
+from tqdm import tqdm
+import urllib 
+import numpy as np
+
+_i = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+os.environ["DBUSERNAME"] = "democracylab"
+os.environ["PW"] = "DemocracyLab2019"
+if _i not in sys.path:
+    # add parent directory to sys.path so utils module is accessible
+    sys.path.insert(0, _i)
+del _i  # clean up global name space
+from shared_code.utils import (
+    check_similarity, locate_potential_duplicate,
+    insert_services, client
+)
+from shared_code.base_scraper import BaseScraper
+
+     
+class Imls_Scraper(BaseScraper):
+  
+    def __init__(self,
+                 source: str,
+                 data_url: str,
+                 data_page_url: str,
+                 data_format: str,
+                 extract_usecols: List[str],
+                 drop_duplicates_columns: List[str],
+                 rename_columns: dict,
+                 service_summary: Any,
+                 check_collection: str,
+                 dump_collection: str,
+                 dupe_collection: str,
+                 data_source_collection_name: str,
+                 collection_dupe_field: str,
+                 encoding: str = 'utf-8'):
+        BaseScraper.__init__(self, source, data_url, data_page_url, data_format, extract_usecols, drop_duplicates_columns, rename_columns, service_summary, check_collection, dump_collection, dupe_collection, data_source_collection_name, collection_dupe_field, encoding)
+        resp = super().scrape_updated_date()
+        soup = BeautifulSoup(resp, 'html.parser')
+        latestLink = ""
+        latestDate = 0
+        for link in soup.findAll('a', href=True, text='CSV'):
+             match = re.match(r'.*([1-3][0-9]{3})', link['href'])
+             if match is not None:
+                year = int(match.group(1))
+                if year > latestDate:
+                    latestDate = year
+                    latestLink = link['href']
+
+        self._data_url: str = latestLink
+        self._latest_date: datetime = datetime.strptime(str(latestDate), '%Y')
+
+    def scrape_updated_date(self) -> str:
+        return self.latest_date.strftime('%Y-%m-%d')
+
+    def grab_data(self) -> pd.DataFrame:      
+        r = requests.get(self.data_url)
+        zf = ZipFile(BytesIO(r.content))
+        for info in zf.infolist():
+            if fnmatch.fnmatch(info.filename, '*_outlet_*.csv'):
+                df = pd.read_csv(zf.open(info.filename),         
+                usecols=self.extract_usecols,
+                encoding = "ISO-8859-1")
+                df = super().grab_data(df=df)
+                df['notes'] = np.where((df['STATSTRU'] == '03') | (df['STATSTRU'] == '23'), 'Closed or Temporarily Closed library.', '')
+                return df
+
+    @property
```

Comment:
We can avoid having this property if we move the logic from init method to scrape_updated_date method.

Reply:
As my previous comment stated, the link to fetch the data is controlled by the logic to check for the updated date. I can repeat the logic in the scrape_updated_method but it made more sense to me to just use a property and reference it later.

Reply:
Right.